optax/losses/_self_supervised_test.py

```diff
@@ -124,6 +124,239 @@ def test_vmap(self, anchor, positive, negative):
+    q2 = jax.random.normal(k2, (2, 3), dtype=jnp.float32)
+    z1 = jax.random.normal(k3, (2, 3), dtype=jnp.float32)
+    z2 = jax.random.normal(k4, (2, 3), dtype=jnp.float32)
+
+    def testing_byol_loss(q1_val, z2_val, q2_val, z1_val, eps=1e-6):
+
+      cos_12 = _regression.cosine_similarity(
+          q1_val, z2_val, epsilon=eps,
+      )
+      cos_21 = _regression.cosine_similarity(
+          q2_val, z1_val, epsilon=eps,
+      )
+      loss_12 = 2.0 - 2.0 * cos_12
+      loss_21 = 2.0 - 2.0 * cos_21
+      loss = 0.5 * (loss_12 + loss_21)
+      return jnp.mean(loss)
+
+    handmade_result = testing_byol_loss(q1, z2, q2, z1)
+    result = jax.jit(_self_supervised.byol_loss)(
+        q1, z2, q2, z1, symmetric=True,
+    )
+    np.testing.assert_allclose(result, handmade_result, atol=1e-4)
+
+  @parameterized.parameters(0, 1, 42)
+  def test_single_direction_matches_handmade(self, seed):
+    key = jax.random.PRNGKey(seed)
+    k1, k2 = jax.random.split(key, 2)
+    q = jax.random.normal(k1, (2, 3), dtype=jnp.float32)
+    z = jax.random.normal(k2, (2, 3), dtype=jnp.float32)
+
+    def testing_single_direction_byol(q_val, z_val, eps=1e-6):
+      cos = _regression.cosine_similarity(
+          q_val, z_val, epsilon=eps,
+      )
+      loss = 2.0 - 2.0 * cos
+      return jnp.mean(loss)
+
+    handmade_result = testing_single_direction_byol(q, z)
+    result = jax.jit(_self_supervised.byol_loss)(
+        q, z, symmetric=False,
+    )
+    np.testing.assert_allclose(result, handmade_result, atol=1e-4)
+
+class SimSiamLossTest(parameterized.TestCase):
+  @parameterized.parameters(0, 1, 42)
+  def test_symmetric_batched_matches_handmade(self, seed):
+    key = jax.random.PRNGKey(seed)
+    k1, k2, k3, k4 = jax.random.split(key, 4)
+    p1 = jax.random.normal(k1, (2, 3), dtype=jnp.float32)
+    p2 = jax.random.normal(k2, (2, 3), dtype=jnp.float32)
+    z1 = jax.random.normal(k3, (2, 3), dtype=jnp.float32)
+    z2 = jax.random.normal(k4, (2, 3), dtype=jnp.float32)
+
+    def testing_simsiam_loss(p1_val, z2_val, p2_val, z1_val, eps=1e-6):
+      cos_12 = _regression.cosine_similarity(
+          p1_val, z2_val, epsilon=eps,
+      )
+      cos_21 = _regression.cosine_similarity(
+          p2_val, z1_val, epsilon=eps,
+      )
+      loss_12 = -cos_12
+      loss_21 = -cos_21
+      loss = 0.5 * (loss_12 + loss_21)
+      return jnp.mean(loss)
+
+    handmade_result = testing_simsiam_loss(p1, z2, p2, z1)
+    result = jax.jit(_self_supervised.simsiam_loss)(
+        p1, z2, p2, z1, symmetric=True,
+    )
+    np.testing.assert_allclose(result, handmade_result, atol=1e-4)
+
+  @parameterized.parameters(0, 1, 42)
+  def test_single_direction_matches_handmade(self, seed):
+    key = jax.random.PRNGKey(seed)
+    k1, k2 = jax.random.split(key, 2)
+    p = jax.random.normal(k1, (2, 3), dtype=jnp.float32)
+    z = jax.random.normal(k2, (2, 3), dtype=jnp.float32)
+
+    def testing_single_direction_simsiam(p_val, z_val, eps=1e-6):
```

Comment:
Let's ideally avoid a re-implementation tests, is there another way to test here?